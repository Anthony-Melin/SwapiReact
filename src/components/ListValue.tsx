import { FC, PropsWithChildren } from 'react'

type ListValueProps = PropsWithChildren<{
    title: string
    item: string
    values: string[]
    Wrap?: FC<PropsWithChildren<{ value: string }>>
    testid: string
}>

const ListValue = ({
    title,
    item,
    values,
    Wrap = ({ value, ...props }) => <p {...props}>{value}</p>,
    testid,
    ...restProps
}: ListValueProps) => {
    return (
        <div data-testid={testid} {...restProps}>
            <p data-testid={`${testid}-title`}>{title}</p>
            {values.length === 0 ? (
                <p data-testid={`${testid}-empty`}>No {item} linked</p>
            ) : (
                <ul>
                    {values.map((value, index) => {
                        return (
                            <li key={`${testid}-${index}`}>
                                <Wrap
                                    data-testid={`${testid}-${index}`}
                                    value={value}
                                />
                            </li>
                        )
                    })}
                </ul>
            )}
        </div>
    )
}

export default ListValue
