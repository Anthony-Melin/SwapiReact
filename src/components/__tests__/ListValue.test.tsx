import { describe, it, expect, beforeEach } from 'vitest'
import { render, screen } from '@testing-library/react'
import ListValue from '../ListValue.tsx'

describe('ListValue', () => {
    describe('with empty list', () => {
        beforeEach(() => {
            render(
                <ListValue
                    title="TITLE"
                    item="ITEM"
                    values={[]}
                    testid="testid"
                />
            )
        })

        it('should match snapshot', () => {
            expect(screen.getByTestId('testid')).toMatchSnapshot()
        })

        it('should have a title', () => {
            expect(screen.getByTestId('testid-title')).toHaveTextContent(
                'TITLE'
            )
        })

        it('should have empty list label', () => {
            expect(screen.getByTestId('testid-empty')).toHaveTextContent(
                'No ITEM linked'
            )
        })
    })

    describe('with filled list', () => {
        beforeEach(() => {
            render(
                <ListValue
                    title="TITLE"
                    item="ITEM"
                    values={['VALUE_1', 'VALUE_2']}
                    testid="testid"
                />
            )
        })

        it('should match snapshot', () => {
            expect(screen.getByTestId('testid')).toMatchSnapshot()
        })

        it('should have a title', () => {
            expect(screen.getByTestId('testid-title')).toHaveTextContent(
                'TITLE'
            )
        })

        it('should contains elements', () => {
            expect(screen.getByTestId('testid-0')).toHaveTextContent('VALUE_1')
            expect(screen.getByTestId('testid-1')).toHaveTextContent('VALUE_2')
        })
    })
})
