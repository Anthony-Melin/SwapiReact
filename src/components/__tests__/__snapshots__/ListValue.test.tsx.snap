// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListValue > with empty list > should match snapshot 1`] = `
<div
  data-testid="testid"
>
  <p
    data-testid="testid-title"
  >
    TITLE
  </p>
  <p
    data-testid="testid-empty"
  >
    No 
    ITEM
     linked
  </p>
</div>
`;

exports[`ListValue > with filled list > should match snapshot 1`] = `
<div
  data-testid="testid"
>
  <p
    data-testid="testid-title"
  >
    TITLE
  </p>
  <ul>
    <li>
      <p
        data-testid="testid-0"
      >
        VALUE_1
      </p>
    </li>
    <li>
      <p
        data-testid="testid-1"
      >
        VALUE_2
      </p>
    </li>
  </ul>
</div>
`;
